lib/src/main/java/motif/ScopeFactory.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (c) 2018-2019 Uber Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package motif;
+
+import motif.internal.Constants;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.HashMap;
+import java.util.Map;
+
+public class ScopeFactory {
+
+    private ScopeFactory() {}
+
+    private static final NoDependencies NO_DEPENDENCIES = new NoDependencies() {};
+
+    private static final Map<Class<?>, Constructor<?>> scopeClassToCreateMethod = new HashMap<>();
+
+    public static <S extends Creatable<NoDependencies>> S create(Class<S> scopeClass) {
+        return create(scopeClass, NO_DEPENDENCIES);
+    }
+
+    public static <S extends Creatable<D>, D> S create(Class<S> scopeClass, D dependencies) {
+        Constructor<?> constructor = getConstructor(scopeClass);
+        try {
+            if (constructor.getParameterTypes().length == 0) {
+                return (S) constructor.newInstance();
+            } else {
+                return (S) constructor.newInstance(dependencies);
+            }
+        } catch (IllegalAccessException e) {
```

Comment:
This would implicitly bump the min-sdk to 19:
![image](https://user-images.githubusercontent.com/847683/70100315-32f80d80-15e6-11ea-917b-d5d2f2d9b0a6.png)

While it may make sense to only support SDK 19 and up, I'd rather make that a conscious decision rather than forcing that limitation to optimize just this piece of code.